Generational ZGC on linux RISCV64

yadongw opened the pull request:
Hi,

Please review this PR, porting zgc generational to RISC-V.

The load barrier we now have for RISC-V:
```
  lui(tmp, barrier_Relocation::unpatched << 12);  // put metadata_bits in tmp[31:12], filling in the lowest 12 bits with zeros
  srli(tmp, tmp, 12);
  andr(tmp, ref, tmp);
  beqz(tmp, good);
  j(*stub->entry());

good:
  srli(ref, ref, ZPointerLoadShift);
```

It's a little bloated, but we can simplified it by introduing bexit from the Zbs extension in the future, and it will look like:
```
  bexti(tmp, ref, barrier_Relocation::unpatched);
  beqz(tmp, good);
  j(*stub->entry());

good:
  srli(ref, ref, ZPointerLoadShift);
```

To go further with the pointer masking from RISC-V "J" extension, we can remove the uncolor instruciton by putting the color bits on the upper 16 bits of z pointers, so that we can have the hardware ignore them at all. 

copy_load_barrier and copy_store_barrier for copy_memory with the risc-v vector extension (-XX:+UseRVV) will be implemented separately.

It includes 2 patches from the jdk mainline that it depends on, one of which is the RISC-V port of loom.

Testing on Linux-riscv64 HiFive Unmatched board:
* Server release & fastdebug build OK.
* Passed tier1 tests (release build).
* Passed jtreg tests under test/hotspot/jtreg/gc/z (release build).
* Perfomed benchmark tests like SPECjvm2008, SPECjbb2015 (release build).
Testing on Linux-riscv64 Qemu:
* Passed jtreg tests under test (release build) without new failures compared to the jdk mainline.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Error
&nbsp;⚠️ This PR only contains changes already present in the target

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/zgc pull/10/head:pull/10` \
`$ git checkout pull/10`

Update a local copy of the PR: \
`$ git checkout pull/10` \
`$ git pull https://git.openjdk.org/zgc pull/10/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10`

View PR using the GUI difftool: \
`$ git pr show -t 10`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/zgc/pull/10.diff">https://git.openjdk.org/zgc/pull/10.diff</a>

</details>

fisk commented:
Thank you for this contribution. I do wonder, why you materialize a shifted load bad mask and then spend an instruction shifting it back down after in the example? Fortunately, looking at the C2 code generation of this patch, I get the impression that you don't actually do that?

yadongw commented:
> Thank you for this contribution. I do wonder, why you materialize a shifted load bad mask and then spend an instruction shifting it back down after in the example? Fortunately, looking at the C2 code generation of this patch, I get the impression that you don't actually do that?

It's a special design in RISC-V. `lui rd, imm` is the only option if you want to load more than 12 bits of immediate value into a register by one single instruction. But it will put the imm[31:12] to rd[31:12], so that a right shift instruction is necessary.
C2 did the same thing, and I used li16u as lui + srli.

fisk commented:
> > Thank you for this contribution. I do wonder, why you materialize a shifted load bad mask and then spend an instruction shifting it back down after in the example? Fortunately, looking at the C2 code generation of this patch, I get the impression that you don't actually do that?
> 
> 
> 
> It's a special design in RISC-V. `lui rd, imm` is the only option if you want to load more than 12 bits of immediate value into a register by one single instruction. But it will put the imm[31:12] to rd[31:12], so that a right shift instruction is necessary.
> 
> C2 did the same thing, and I used li16u as lui + srli.

Ah, I see. Well the load barriers do look a bit painful, but with that extension it looks like it will be better. We can start like this and then improve later on. Anyway, perhaps we should rebase generational ZGC first so we can cut off the loom stuff from this change?

yadongw commented:
> > > Thank you for this contribution. I do wonder, why you materialize a shifted load bad mask and then spend an instruction shifting it back down after in the example? Fortunately, looking at the C2 code generation of this patch, I get the impression that you don't actually do that?
> > 
> > 
> > It's a special design in RISC-V. `lui rd, imm` is the only option if you want to load more than 12 bits of immediate value into a register by one single instruction. But it will put the imm[31:12] to rd[31:12], so that a right shift instruction is necessary.
> > C2 did the same thing, and I used li16u as lui + srli.
> 
> Ah, I see. Well the load barriers do look a bit painful, but with that extension it looks like it will be better. We can start like this and then improve later on. Anyway, perhaps we should rebase generational ZGC first so we can cut off the loom stuff from this change?

Yeah, it'd better to rebase first, after then I can make a clean patch.

xiangzhai commented on src/hotspot/cpu/riscv/gc/z/z_riscv64.ad:
```diff
@@ -31,169 +31,218 @@ source_hpp %{
+    Address ref_addr($mem$$Register);
+    z_color(_masm, this, $oldval_tmp$$Register, $oldval$$Register, t0);
+    z_store_barrier(_masm, this, ref_addr, $newval$$Register, $newval_tmp$$Register, t1, true /* is_atomic */);
+    __ cmpxchg($mem$$Register, $newval$$Register, $newval_tmp$$Register, Assembler::int64, Assembler::aq /* acquire */, Assembler::rl /* release */, $res$$Register);
```
oldval_tmp?

```
__ cmpxchg($mem$$Register, $newval$$Register /* oldval_tmp */, $newval_tmp$$Register, Assembler::int64, Assembler::aq /* acquire */, Assembler::rl /* release */, $res$$Register);
```

xiangzhai commented on src/hotspot/cpu/riscv/gc/z/zBarrierSetAssembler_riscv.cpp:
```diff
@@ -54,14 +57,305 @@
+      // Stores on relocatable objects never need to deal with raw NULL pointers in fields.
+      // Raw NULL pointers may only exists in the young generation, as they get pruned when
+      // the object is relocated to old. And no pre-write barrier needs to perform any action
+      // in the yound generation.
```
young.

xiangzhai commented on src/hotspot/cpu/riscv/gc/z/zBarrierSetAssembler_riscv.cpp:
```diff
@@ -72,70 +366,266 @@ void ZBarrierSetAssembler::store_at(MacroAssembler* masm,
+
+    __ enter();
+    if (_result != noreg) {
+      __ push_call_clobbered_registers_except(_result);
```
RegSet::of(_result)

xiangzhai commented on src/hotspot/cpu/riscv/gc/z/zBarrierSetAssembler_riscv.cpp:
```diff
@@ -72,70 +366,266 @@ void ZBarrierSetAssembler::store_at(MacroAssembler* masm,
+      if (_result != x10) {
+        __ mv(_result, x10);
+      }
+      __ pop_call_clobbered_registers_except(_result);
```
Ditto

fisk commented:
Hi guys. We have rebased now upon the latest jdk/master, so if you rebase this patch now, we can integrate the RISC-V changes.

RealFYang commented:
@yadongw : Hi, you might want to update the copyright years for the RISC-V specific files changed. Thanks.

yadongw committed:
implement zgc generational on riscv

yadongw commented:
> Hi guys. We have rebased now upon the latest jdk/master, so if you rebase this patch now, we can integrate the RISC-V changes.

I have rebased the riscv changes, tested on both hardware and QEMU, and I will push force to update this PR soon. It's a simple approach, although it will lose the review history. I think it's also OK, after all, not the main line.

RealFYang commented:
Also performed tier1-4 tests on linux-aarch64 platform, result looks good.

PS: Witnessed some issues running SPECjvm2008 startup with fastdebug build, please check.

yadongw commented:
> Also performed tier1-4 tests on linux-aarch64 platform, result looks good.

I see some failures to be fixed and more tests need to be done.

yadongw committed:
fix unreachable branches in debug mode

yadongw committed:
fix zoffset overflow when allocating large objects

yadongw commented:
> Also performed tier1-4 tests on linux-aarch64 platform, result looks good.
> 
> PS: Witnessed some issues running SPECjvm2008 startup with fastdebug build, please check.

Thanks for the test. I fixed them in a work-around way and tested test/hotspot/jtreg/gc/z with fastdebug build. 

The reason for the failures is that a zoffset greater than or equal to ZAddressOffsetMax will be generated as new_top by adding the current top with size in ZPage::alloc_object(size_t size), which breaks the assertion that zoffset must be less than or equal to ZAddressOffsetMask.

@fisk  How do you think about it?

fisk commented:
Thank you for the RISCV port @yadongw! The only thing I have some thoughts about is the zoffset range checking code. But it seems to be a pre-existing issue. I'm okay with taking in what you have, and then I'll see how the zoffset range checking code can be massaged a bit after. Does that sound good to you?

yadongw commented:
> Thank you for the RISCV port @yadongw! The only thing I have some thoughts about is the zoffset range checking code. But it seems to be a pre-existing issue. I'm okay with taking in what you have, and then I'll see how the zoffset range checking code can be massaged a bit after. Does that sound good to you?

@fisk Hi, Erik. That sounds good to me. I can type /integrate if you approve this PR.
